### PR TITLE
Add Rebar 3 support.

### PR DESCRIPTION
**Pull request by russss**
The magic to compile NIFs has been removed from Rebar 3 - it's now
[recommended to create a Makefile](http://www.rebar3.org/v3.0/docs/building-cc). These changes do that and are
compatible with Rebar 2 and 3.

### CODE REVIEW

**Comment by ypaq** [comment]
:+1:

**Comment by jlouis** [comment]
+1, I'd pay to get this too :)

**Comment by Vagabond** [comment]
Is this Makefile also compatible with BSD make?

**Comment by jlouis** [comment]
Excellent question! I have a FreeBSD machine here, so I'll test that it works. I've tested OSX and Linux for the moment, but I'll take FreeBSD/BSD-make into account. It may not be compatible, but we'll see.

**Comment by jlouis** [comment]
Verdict: The makefile is not compatible with BSD make, but you can get around that by installing GNU Make (which defaults itself to gmake on FreeBSD) and then use the following `rebar.config` diff:

```
diff --git a/rebar.config b/rebar.config
index 6a18fb2..53ef5b4 100644
--- a/rebar.config
+++ b/rebar.config
@@ -1,5 +1,9 @@
 {erl_opts, [debug_info]}.
 {deps, []}.

-{pre_hooks, [{"(linux|darwin|solaris)", compile, "make -C c_src"}]}.
-{post_hooks, [{"(linux|darwin|solaris)", clean, "make -C c_src clean"}]}.
+{pre_hooks,
+     [{"(linux|darwin|solaris)", compile, "make -C c_src"},
+      {"(freebsd)", compile, "gmake -C c_src"}]}.
+{post_hooks,
+     [{"(linux|darwin|solaris)", clean, "make -C c_src clean"},
+      {"(freebsd)", clean, "gmake -C c_src clean"}]}.
```

**Comment by Vagabond** [comment]
The other option is to have a Makefile and a GNUMakefile, GNU make will read the GNUMakefile first, but BSD make will read Makefile. Then you can have the tasks in the Makefile do 'gmake <task>'

An example can be found here:

https://github.com/basho/erlang_js/blob/develop/Makefile
https://github.com/basho/erlang_js/blob/develop/GNUmakefile

**Comment by jlouis** [comment]
If you were to make that switch, I'd much rather write a working BSD Makefile and use that for GNU Make as well. It seems wrong to maintain two makefiles for something as simple as compiling a C NIF.

File: c_src/Makefile [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	CC ?= gcc
+	CFLAGS ?= -O3 -std=c99 -finline-functions -Wall -Wmissing-prototypes
+	CXXFLAGS ?= -O3 -finline-functions -Wall
+endif
+
+CFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
+CXXFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
+
+LDLIBS += -L $(ERL_INTERFACE_LIB_DIR) -lerl_interface -lei
+LDFLAGS += -shared
```
Comment (goofansu) [inline c_src/Makefile]: ` -undefined dynamic_lookup` should be appended here to resolve `ld: symbol(s) not found for architecture x86_64` error.

The full error information:

```
~/g/erlang-syslog ❯❯❯ rebar3 compile
CXXFLAGS += -fPIC -I $(ERTS_INCLUDE_DIR) -I $(ERL_INTERFACE_INCLUDE_DIR)
===> Verifying dependencies...
===> Compiling syslog
cc /Users/suyejun/github/erlang-syslog/c_src/syslog_drv.o -L/usr/local/lib -shared -L /usr/local/Cellar/erlang/18.2.1/lib/erlang/lib/erl_interface-3.8.1/lib -lerl_interface -lei -o /Users/suyejun/github/erlang-syslog/c_src/../priv/syslog_drv.so
Undefined symbols for architecture x86_64:
  "_driver_alloc", referenced from:
      _syslogdrv_start in syslog_drv.o
      _syslogdrv_control in syslog_drv.o
  "_driver_free", referenced from:
      _syslogdrv_stop in syslog_drv.o
      _syslogdrv_control in syslog_drv.o
  "_set_port_control_flags", referenced from:
      _syslogdrv_start in syslog_drv.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [/Users/suyejun/github/erlang-syslog/c_src/../priv/syslog_drv.so] Error 1
```

**Comment by binarytemple-bet365** [comment]
Fixed by my P/R for rebar3 upgrade. Suggest close

**Pull request closed without merging**